rfc/languages-in-w3s.md

```diff
@@ -0,0 +1,84 @@
+- Every hiring manager loves javascript cause it's easy to hire javascript developers. The main problem with this approach is there are so many kinds of javascript development and ways to write javascript that hiring a javascript developer is not a gaurantee they'll be able to write javascript the way your team uses it.
+
+#### Some Notes On Our Javascript
+
+- Thankfully this team has unified on a coding style, linter, tool chain, package manager, and approach to typing. +1 on that.
+- We have not however documented why we made the choices we did. We should do so ASAP. In particular types in JSDoc style comments and why we do it needs documentation. This includes our sporadic dropping into Typescript and why we do it. (For example: 'why is there an api.js and an api.ts and the js file just says `export {}` -- the answer is 'cause modules' but is EXTREMELY not clear)
+- The w3up repo is huge, and honestly, even if you know javascript, there's a lot to get up to speed on. We should have a guide to the code.
+
+### Go
+
+- Go is a highly performant garbage collected language like Java, but is much simpler than Java
+- Go has an extremely powerful concurrency model with green threads
+- These two components together make Go one of the most powerful languages for writing servers of any kind. The challenge of server programming is not super fast compute, but super high concurrency execution. Go excels at this, which matters cause our project has a lot of servers involved.
+- Go's ability to do binary distributions by default is a major plus for kind of project.
+- By and large, people get stuff done quickly in Go because Go is extremely simple. It's very fast to learn, easy to read, and it all looks the same. The only problem is its extremely boring. (Note from Hannah: When I first started writing Go I hated it. I thought someone took the joy out of programming. Now I love it cause I get my joy from what I'm building, not my programming language)
+- Lots and lots of Filecoin tooling and software is written in Go. In fact, W3S already has Go in it, we just don't maintain it -- it's called Spade
+- Garbage collection and green threads present challenges for using Go over an FFI, or for compiling go to WebAssembly, though TinyGo has improved this story by reducing the runtime size.
+
+### Rust
+
+Caveat from Hannah: I am not a highly experienced Rust dev
+
+- Rust is a better C/C++. Like definitely much better. This is huge cause no one wrote a better C/C++ for low level programming for 20 years.
+- However, by and large, you should probably not write programs in Rust unless you'd consider writing them in C/C++. There is minimal evidence one can be as productive in Rust as higher level languages.
```

Comment:
I'm not sure what the basis of this assessment is ? Rust can be highly productive high level language as long as you grok resource ownership model. I would even argue that forcing you to think about ownership is a good thing and a perfect pairing with systems utilizing object capability model (like UCANs) because it's all about who owns the resource and who is given access at time.

Reply:
I also agree that really depends on what you're trying to achieve. When I was first learning Rust, I rewrote a bunch of my personal organization scripts in Rust and it felt quite productive.

Reply:
Ultimately, I'm interested not in the nice to haves but the must haves. Resource ownership is a great thing to understand. I take care to not pass around a lot of mutable data in whatever language I'm working in, and I try to have a pretty clear chain of provenance for data structures in my head. But do you need the strict resource ownership of a borrow checker for your specific program?

To me the answer is "only when absolute performance is a must". Cause there is a non-zero overhead to working without garbage collection.

The driver I want to keep coming back to is "what will get us to the end of our roadmap fastest". How fast can you write a high performance web server in Rust? Especially if you're building a new service from scratch? (many of the Rust success stories are often rewrites of an existing hot path service for performance reasons -- which makes sense when you reach the point optimization is no longer premature)

Reply:
It depends on a bunch of factors.

- In the FVM, we had some issues because wasmtime needed a _static_ reference to some state and we didn't want to have rc/refcells all over the place. But, once we got past that issue, it wasn't too bad.
- In the builtin actors, we had a couple of issues translating existing go idioms, but wouldn't have had any issues if we had written them in rust first.
- If you dig too deep into the type system, there be Balrogs. You _will_ find things that really should compile but don't because rust just can't quite reason about your types.

The biggest drawbacks I see are:

1. Compile times and test time. The former is generally an issue if you're have a complex build pipeline with lots of deriving, macros, generics, etc. Especially if you have many build targets (e.g., the builtin actors). The latter can be an issue because, e.g., if you test with an unoptimized build (to improve compile times), tests can take a while in some cases (was also a problem in the builtin actors, fixed by compiling _specific_ crates with optimizations regardless of the build target).
2. Learning curve. This is something people get over, but it _does_ take a little longer to onboard new teammates.
3. **TEMPTATION**. The type system is _very_ expressive and it's easy to go down rabbit holes. This is one of the _biggest_ dangers I've seen in rust because there's always a good reason and the issue is only obvious in hindsight. Really, trying to find the perfect abstraction is probably the biggest rust time sink.
4. Async/await. IMO, it's still not ready (from what I've seen). It's getting better every release, but it still gets in the way enough to slow development down.

On the other hand, I absolutely love rust and write most of my personal projects in rust (those not written in Elisp). Mostly because:

1. Performance is very consistent. E.g., when we switched from the "legacy" VM to the FVM, block times, GC spikes, etc. got _much_ more consistent in lotus. A lot of this is simply because rust forces you to think about resource allocation and lifecycles, where as go requires some more discipline.
2. The type system is very expressive. We've had very few bugs in the FVM because we can often statically prevent them. But you have to be careful to not waste too much time trying to statically prevent all bugs.
3. The meta-programming and operator overloading allow for some fun patterns _when_ you have common patterns you need to abstract away (they can also be a time sink if you put too much effort into them up-front).

I actually do write high-level tools in rust if I need them to just work and it generally doesn't take that long. There are lots of really great high-level libraries for writing CLI tools. I used to write them in python, but then getting them to work reliably always took a while (need to write tests, etc.).

Rust only starts to slow you down when you get into more complex projects and start building out complex abstractions. And, TBH, a lot of the trouble is self-inflicted because rust _lets_ you build those abstractions.

Reply:
^^ @Stebalien thanks for taking the time 

Folks I think this is worth reading

Reply:
> * In the builtin actors, we had a couple of issues translating existing go idioms, but wouldn't have had any issues if we had written them in rust first.

I actually was going to comment something along those lines but restrained myself. Going from TS to Rust would feel a lot more at home than going from TS to Go or from Go to Rust. In fact most of our TS actually feels more like Rust than JS because we put a lot of care into resource management to keep memory use at desired bounds.

> TEMPTATION. The type system is very expressive and it's easy to go down rabbit holes. This is one of the biggest dangers I've seen in rust because there's always a good reason and the issue is only obvious in hindsight. Really, trying to find the perfect abstraction is probably the biggest rust time sink.

I could not agree more! Most typed languages (I would not put go into this group) it's really easy to fall in the trap of trying to increase safety by capturing more in the type system. I find Rust and TS both to be particularly luring there, but then again solution we often resort to is: Dont' abstract until you have an actual use case for that abstraction, or put it differently if you have one implementation of the abstraction that abstraction is just a means of indirection and unnecessary complexity.

Reply:
Also, don't spend a bunch of time re-writing your abstractions. Yes, I'd agree that TS is similar in that respect (and not in a good way).